linux/devdoc/threadpool_linux_requirements.md

```diff
@@ -0,0 +1,258 @@
+3. `semaphore` : a semaphore to ensure the number of waiting tasks changed automically when a task is inserted or consumed in the array.
+4. `task_array` : a circular array of waiting tasks with default size 2048, initialized in `threadpool_create`. Each array item contains a `THREADPOOL_TASK` object, which obtains the `task_state` and task function information.
+    - `task_state` can be following:
+       - `TASK_NOT_USED` : entered after a call to `threadpool_create`, new array items after a call to `reallocate_threadpool_array`, and a call to `threadpool_work_func`.
+       - `TASK_INITIALIZING` : entered in the progress of a call to `threadpool_schedule_work`.
+       - `TASK_WAITING` : entered after a call to `threadpool_schedule_work` with a zero return value.
+       - `TASK_WORKING` : entered in the progress of a call to `threadpool_work_func` by calling `threadpool_open_async`.
+5. `task_array_size` : a 32 bit variable represents the size of the task array.
+6. `insert_idx` : a 64 bit variable represents the next available insert position in the circular task array.
+7. `consume_idx` : a 64 bit variable represents the next available consume position in the circular task array.
+8. `pending_call_context` : a 64 bit variable represents the tasks waiting to be scheduled. The value is initialized to be 0 in `threadpool_create`, incremented when a new task is inserting into the task array in `threadpool_schedule_work`, and decremented after the task has been scheduled.
+9. `srw_lock` : a SRW lock to ensure the progress of resize array will not be interupted as well as task array insert and consume get locked when task array resize happens. 
+10. `thread_handle_array` : an array of threads in the `threadpool`, all intialized in `threadpool_open_async`. 
+
+### Reallocating the Task Array
+
+When there are no unused slots in the circular task array, the task array size is doubled by calling `reallocate_threadpool_array`. New task array memory layout should be in same order of the original array with emilinated gaps between items.  `compress_count` is used to record the number of gaps between `insert_idx` and `consume_idx`. New task array should initialize `consume_idx` as -1 and `insert_idx` as (original array size - gap size - 1) to get the correct index when incremented. If the doubled array size get overflowed for a 32 bits integer, resize failed and return. In this case, the maximum size of task array is 2^32 and the maximum value of `insert_idx` and `consume_idx` is 2^64 which will never get overflowed.
+
+```mermaid
+graph TD
+    A(insert_idx == consume_idx) --> |worker thread runs first| B(consume_idx increment)
+    A--> |main thread runs first| D(no gaps) -->F
+    B--> C(gaps occured)-->E(memmove)-->F(resize done)
+    
+```
+
+## Exposed API
+```C
+
+typedef struct THREADPOOL_TAG* THREADPOOL_HANDLE;
+typedef struct TIMER_INSTANCE_TAG* TIMER_INSTANCE_HANDLE;
```

Comment:
>TIMER_INSTANCE_HANDLE [](http://example.com/codeflow?start=35&length=21)

This is the first time I see in the .md a reference to timers. There is nothing in the overview or design section about timers. If this module handles timers too, we should add the info about it above.

Reply:
added in overview and function details below